src/ompl/geometric/planners/experience/ThunderRetrieveRepair.h

```diff
@@ -0,0 +1,205 @@
+#ifndef OMPL_TOOLS_THUNDER_THUNDER_RETRIEVE_REPAIR_
+#define OMPL_TOOLS_THUNDER_THUNDER_RETRIEVE_REPAIR_
+
+#include <ompl/geometric/planners/PlannerIncludes.h>
+#include <ompl/geometric/PathGeometric.h>
+#include <ompl/geometric/PathSimplifier.h>
+#include <ompl/datastructures/NearestNeighbors.h>
+
+namespace ompl
+{
+    namespace tools
+    {
+        OMPL_CLASS_FORWARD(ThunderDB);
+    }
+
+    namespace geometric
+    {
+
+        /// @cond IGNORE
+        /** \brief Forward declaration of ompl::base::ThunderRetrieveRepair */
+        OMPL_CLASS_FORWARD(ThunderRetrieveRepair);
+        /// @endcond
+
+        /** \class ompl::base::ThunderRetrieveRepairPtr
+            \brief A boost shared pointer wrapper for ompl::base::ThunderRetrieveRepair */
+
+        /**
+           @anchor ThunderRetrieveRepair
+           @par Short description
+           Thunder is an experience-based planning framework that learns to reduce computation time
+           required to solve high-dimensional planning problems in varying environments.
+           @par External documentation
+           Berenson, Dmitry, Pieter Abbeel, and Ken Goldberg: A robot path planning framework that learns from experience, in <em>Robotics and Automation (ICRA), 2012 IEEE International Conference on. IEEE</em>, 2012.
+           David Coleman, Ioan A. Sucan, Mark Moll, Kei Okada, Nikolaus Correll, "Experience-Based Planning with Sparse Roadmap Spanners"
+           <a href="http://arxiv.org/pdf/1410.1950.pdf">[PDF]</a>
+        */
+
+        /** \brief The Thunder Framework's Retrieve-Repair component */
+        class ThunderRetrieveRepair : public base::Planner
+        {
+        public:
+
+            /** \brief Constructor */
+            ThunderRetrieveRepair(const base::SpaceInformationPtr &si, const ompl::tools::ThunderDBPtr &experienceDB);
+
+            virtual ~ThunderRetrieveRepair(void);
+
+            /** \brief Get information about the exploration data structure the planning from scratch motion planner used. */
+            virtual void getPlannerData(base::PlannerData &data) const;
+
+            /**
+             *  \brief Get debug information about the top recalled paths that were chosen for further filtering
+             *  \return data - vector of PlannerData objects that each hold a single path
+             */
+            const std::vector<ompl::geometric::PathGeometric>& getLastRecalledNearestPaths() const;
+
+            /**
+             *  \brief Get debug information about the top recalled paths that were chosen for further filtering
+             *  \return chosenID - the index of the PlannerData object that was chosen for repair
+             */
+            const std::size_t& getLastRecalledNearestPathChosen() const;
+
+            /**
+             * \brief Get the chosen path used from database for repair
+             * \return PlannerData of chosen path
+             */
+            ompl::geometric::PathGeometric getChosenRecallPath() const;
```

Comment:
No way to return a const & here? If not, maybe call this copyChosenRecallPath(geometric::PathGeometric *copy) ?